Source/WebKit/GPUProcess/graphics/wc/WCRemoteFrameHostLayerManager.cpp

```diff
@@ -0,0 +1,111 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WCRemoteFrameHostLayerManager.h"
+
+#if USE(GRAPHICS_LAYER_WC)
+
+#include <WebCore/BitmapTexture.h>
+#include <WebCore/TextureMapperPlatformLayer.h>
+#include <wtf/NeverDestroyed.h>
+
+namespace WebKit {
+
+class WCRemoteFrameHostLayer final : public WebCore::TextureMapperPlatformLayer {
+    WTF_MAKE_FAST_ALLOCATED();
+public:
+    void paintToTextureMapper(WebCore::TextureMapper& textureMapper, const WebCore::FloatRect& targetRect, const WebCore::TransformationMatrix& modelViewMatrix, float opacity) override
+    {
+        if (m_texture)
+            textureMapper.drawTexture(*m_texture, targetRect, modelViewMatrix, opacity);
+    }
+
+    void setTexture(RefPtr<WebCore::BitmapTexture> texture)
+    {
+        m_texture = texture;
+    }
+
+private:
+    RefPtr<WebCore::BitmapTexture> m_texture;
+};
+
+class WCRemoteFrameHostLayerManager::RemoteFrameHostLayerData final {
+    WTF_MAKE_FAST_ALLOCATED();
+public:
+    RemoteFrameHostLayerData(WebCore::ProcessIdentifier webProcessIdentifier)
+        : m_webProcessIdentifier(webProcessIdentifier)
+    {
+    }
+
+    WebCore::ProcessIdentifier webProcessIdentifier() { return m_webProcessIdentifier; }
+    void setWebProcessIdentifier(WebCore::ProcessIdentifier webProcessIdentifier) { m_webProcessIdentifier = webProcessIdentifier; }
+    WCRemoteFrameHostLayer& layer() { return m_layer; }
+
+private:
+    // A web process identifier currently owning the texture. Remote frame process or frame host process.
+    WebCore::ProcessIdentifier m_webProcessIdentifier;
+    WCRemoteFrameHostLayer m_layer;
+};
+
+WCRemoteFrameHostLayerManager& WCRemoteFrameHostLayerManager::singleton()
+{
+    static NeverDestroyed<WCRemoteFrameHostLayerManager> manager;
+    return manager;
+}
+
+WebCore::TextureMapperPlatformLayer* WCRemoteFrameHostLayerManager::acquireRemoteFrameHostLayer(WebCore::LayerHostingContextIdentifier layerHostingContextIdentifier, WebCore::ProcessIdentifier webProcessIdentifier)
+{
+    auto& data = *m_layers.ensure(layerHostingContextIdentifier, [&] {
+        return makeUnique<RemoteFrameHostLayerData>(webProcessIdentifier);
+    }).iterator->value;
+    // Transfer the data ownership to the frame host process.
+    data.setWebProcessIdentifier(webProcessIdentifier);
+    return &data.layer();
+}
+
+void WCRemoteFrameHostLayerManager::releaseRemoteFrameHostLayer(WebCore::LayerHostingContextIdentifier layerHostingContextIdentifier)
+{
+    m_layers.remove(layerHostingContextIdentifier);
+}
+
+void WCRemoteFrameHostLayerManager::updateTexture(WebCore::LayerHostingContextIdentifier layerHostingContextIdentifier, WebCore::ProcessIdentifier webProcessIdentifier, RefPtr<WebCore::BitmapTexture> texture)
```

Comment:
RefPtr<WebCore::BitmapTexture>&& might reduce some ref churn.

Reply:
I use `WTFMove` both inside and callsite of updateTexture. There should be no ref-churn here.